### PR TITLE
IPv6 attack tests

### PR DESCRIPTION
**Pull request by gpotter2**

### CODE REVIEW

File: test/regression.uts [hunk 1]
```diff
@@ -3558,6 +3558,91 @@ tr6.graphdef.startswith("digraph trace {") == True
...
+import mock
+
+@mock.patch("scapy.layers.inet6.sniff")
+@mock.patch("scapy.layers.inet6.sendp")
+def test_attack(function, pktlist, sendp_mock, sniff_mock, options=()):
+    pktlist = [Ether(raw(x)) for x in pktlist]
+    ret_list = []
+    def _fake_sniff(lfilter=None, prn=None, **kwargs):
+        for p in pktlist:
+            if lfilter(p):
```
Comment (guedou) [inline test/regression.uts]: `if lfilter and lfilter(p) and prn:`

File: test/regression.uts [hunk 2]
```diff
@@ -3558,6 +3558,91 @@ tr6.graphdef.startswith("digraph trace {") == True
...
+                prn(p)
+    sniff_mock.side_effect = _fake_sniff
+    def _fake_sendp(pkt, *args, **kwargs):
+        ret_list.append(pkt)
```
Comment (guedou) [inline test/regression.uts]: I think that it will be easier to handle if you build a list of dissected packets.

File: test/regression.uts [hunk 3]
```diff
@@ -3558,6 +3558,91 @@ tr6.graphdef.startswith("digraph trace {") == True
...
+    return ret_list
+
+= Test NDP_Attack_DAD_DoS_via_NS
+
+data = [Ether(src='aa:aa:aa:aa:aa:aa', dst='bb:bb:bb:bb:bb:bb')/IPv6(src="::", dst="ff02::1:ff00:1111")/ICMPv6ND_NS(tgt="ffff::1111", code=17, res=3758096385),
+        Ether(src='aa:aa:aa:aa:aa:aa', dst='ee:ee:ee:ee:ee:ee')/IPv6(src="::", dst="ff02::1:ff5d:c353")/ICMPv6ND_NS(tgt="b643:44c3:f659:f8e6:31c0:6437:825d:c353"),
+        Ether()/IP()/ICMP()]
+results = test_attack(NDP_Attack_DAD_DoS_via_NS, data)
+assert len(results) == 2
+assert raw(results[0]).endswith(b'`\x00\x00\x00\x00\x18:\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\xff\x00\x11\x11\x87\x00X\x86\x00\x00\x00\x00\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x11\x11')
```
Comment (guedou) [inline test/regression.uts]: That will be better to use Scapy object instead of strings here.

File: test/regression.uts (hunk 2 above)
Comment (gpotter2) [reply test/regression.uts]: Isn't it already the case ? pkt is a `Packet` object passed to `sendp`

File: test/regression.uts [hunk 4]
```diff
@@ -3558,6 +3558,93 @@ tr6.graphdef.startswith("digraph trace {") == True
...
+    return ret_list
+
+= Test NDP_Attack_DAD_DoS_via_NS
+
+data = [Ether(src='aa:aa:aa:aa:aa:aa', dst='bb:bb:bb:bb:bb:bb')/IPv6(src="::", dst="ff02::1:ff00:1111")/ICMPv6ND_NS(tgt="ffff::1111", code=17, res=3758096385),
+        Ether(src='aa:aa:aa:aa:aa:aa', dst='ee:ee:ee:ee:ee:ee')/IPv6(src="::", dst="ff02::1:ff5d:c353")/ICMPv6ND_NS(tgt="b643:44c3:f659:f8e6:31c0:6437:825d:c353"),
+        Ether()/IP()/ICMP()]
+results = test_attack(NDP_Attack_DAD_DoS_via_NS, data)
+assert len(results) == 2
+assert results[0][IPv6] == IPv6(b'`\x00\x00\x00\x00\x18:\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\xff\x00\x11\x11\x87\x00X\x86\x00\x00\x00\x00\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x11\x11')
```
Comment (guedou) [inline test/regression.uts]: Sorry, I meant to check the results as done [here](L2029).

Comment (gpotter2) [reply test/regression.uts]: Updated

File: test/regression.uts [hunk 5]
```diff
@@ -3558,6 +3558,173 @@ tr6.graphdef.startswith("digraph trace {") == True
...
+assert b[IPv6].src == "b643:44c3:f659:f8e6:31c0:6437:825d:c353"
+assert b[IPv6].dst == "ff02::1:ff5d:c353"
+assert b[IPv6].hlim == 255
+assert b[ICMPv6ND_NA].tgt == "b643:44c3:f659:f8e6:31c0:6437:825d:c353"
+assert b[ICMPv6NDOptDstLLAddr].lladdr == "ab:ab:ab:ab:ab:ab"
+
+= Test NDP_Attack_NA_Spoofing
+
+data = [Ether(src='aa:aa:aa:aa:aa:aa', dst='bb:bb:bb:bb:bb:bb')/IPv6(src="753a:727c:97b5:f71d:51ea:3901:ab52:e110", dst="ff02::1:ffd4:e5f6")/ICMPv6ND_NS(tgt="ff02::1:ffd4:e5f6", code=171, res=3758096),
+        Ether(src='aa:aa:aa:aa:aa:aa', dst='ee:ee:ee:ee:ee:ee')/IPv6(src="753a:727c:97b5:f71d:51ea:3901:ab52:e110", dst="fe9c:98b0:52b5:7033:5db0:394f:e468:c94f")/ICMPv6ND_NS(),
```
Comment (guedou) [inline test/regression.uts]: I don't understand why this packet works as it is failing the test at line 3540 as the destination address is invalid.

File: test/regression.uts [hunk 6]
```diff
@@ -3558,6 +3558,173 @@ tr6.graphdef.startswith("digraph trace {") == True
...
+    sniff_mock.side_effect = _fake_sniff
+    def _fake_sendp(pkt, *args, **kwargs):
+        ret_list.append(Ether(raw(pkt)))
+    sendp_mock.side_effect = _fake_sendp
+    function(*options)
+    return ret_list
+
+= Test NDP_Attack_DAD_DoS_via_NS
+
+data = [Ether(src='aa:aa:aa:aa:aa:aa', dst='bb:bb:bb:bb:bb:bb')/IPv6(src="::", dst="ff02::1:ff00:1111")/ICMPv6ND_NS(tgt="ffff::1111", code=17, res=3758096385),
```
Comment (guedou) [inline test/regression.uts]: Ether.src should be:
```
In [2]: in6_getnsmac(inet_pton(socket.AF_INET6, "ff02::1:ff00:1111"))
Out[2]: '33:33:ff:00:11:11'
```

**Comment by gpotter2** [comment]
I tried to update it accordingly to your comments.
If I missed something, please let me know

**Comment by guedou** [comment]
Thanks your are awesome! I will merge it when the tests finish.

**Comment by p-l-** [comment]
I've restarted an error test (not failing, error).

**Pull request merged**